cms/search/publishers.py

```diff
@@ -0,0 +1,166 @@
+
+
+class BasePublisher(ABC):
+    """BasePublisher defines shared functionalities, such as how to build a message
+    for created/updated or deleted events. Each subclass only needs to define how
+    to actually publish (i.e., send) the built message.
+
+    dp-search-data-extractor spec link:
+    https://github.com/ONSdigital/dp-search-data-extractor/blob/develop/specification.yml#L53
+    dp-search-data-importer spec link:
+    https://github.com/ONSdigital/dp-search-data-importer/blob/30fb507e90f2cf1974ec0ca43bb0466307e2f112/specification.yml#L186
+    contract: https://github.com/ONSdigital/dis-search-upstream-stub/blob/main/docs/contract/resource_metadata.yml
+    """
+
+    def publish_created_or_updated(self, page: "Page") -> None:
+        """Build the message for the created/updated event.
+        Delegate sending to the subclass's _publish().
+        """
+        channel = self.get_created_or_updated_channel()
+        message = self._construct_message_for_create_update(page)
+        return self._publish(channel, message)
+
+    def publish_deleted(self, page: "Page") -> None:
+        """Build the message for the deleted event.
+        Delegate sending to the subclass's _publish().
+        """
+        channel = self.get_deleted_channel()
+        message = {
+            "uri": page.url_path,
+        }
+        return self._publish(channel, message)
+
+    @abstractmethod
+    def _publish(self, channel: str | None, message: dict) -> None:
+        """Each child class defines how to actually send/publish
+        the message (e.g., Kafka, logging, etc.).
+        """
+
+    @abstractmethod
+    def get_created_or_updated_channel(self) -> str | None:
+        """Provide the channel (or other necessary routing key) for created/updated.
+        This can be a no-op or empty string for some implementations.
+        """
+
+    @abstractmethod
+    def get_deleted_channel(self) -> str | None:
+        """Provide the channel (or other necessary routing key) for deleted messages.
+        This can be a no-op or empty string for some implementations.
+        """
+
+    def _construct_message_for_create_update(self, page: "Page") -> dict:
+        """Build a dict that matches the agreed metadata schema for 'created/updated'.
+
+        resource_metadata.yml:
+           - StandardPayload: requires {uri, title, content_type}, optional fields
+           - ReleasePayload: extends StandardPayload with release-specific fields
+        """
+        # Common fields for StandardPayload (also part of ReleasePayload)
+        # The schema requires at minimum: uri, title, content_type
+        message = {
+            "uri": page.url_path,
+            "content_type": page.search_index_content_type,
+            "release_date": (
+                page.release_date.isoformat()
+                if page.search_index_content_type == "release" and getattr(page, "release_date", None)
+                else None
+            ),
```

Comment:
1. If the `release_date` is `None`, then you need to add the property `provisional_date` and populate that with the value from `release_date_text`. Currently, this is only done for `release` types therefore it should be done in `_construct_release_specific_fields`.
2. What implied this field is only for type release?. I can't see the spec suggesting this (we can resolve in a follow-up PR to get things moving, but I think we need to populate release date for articles etc - have asked for confirmation (lets follow up))

Reply:
Another 🦅 👁️! 💪🏼 
Alas, I intepreted `release_date` as being relevant only to release calendar, but the spec is clear it is in the main corpus https://github.com/ONSdigital/dis-search-upstream-stub/blob/main/docs/contract/resource_metadata.yml#L58

Reply:
I forgot we had the `release_date` field on the article, so that's where this should populate from for that page. Worth resolving here:
- Articles/Calendar Entry - `release_date` field
- Methodology - `publication_date` field

Reply:
Yes, I went `spec` blind at this point, working on this haha. Correlated `release_date` with `release_calendar`, made an assumption, similar to @zerolab , even though `spec` says otherwise.

`Methodology - publication_date field (could create an alias release_date property to simplify logic)` @zerolab thankfully has already done this 2 months ago, so this works seamlessly!

I made it so we look for the `release_date` for all the pages we care about, if they don't have a `release_date` then we set `release_date` to `None`.

If the page type is `release` then it either has to have a `release_date` or `provisional_date`.

Done!

Reply:
Done!